### PR TITLE
Add c2d QoS2 scenario to DeveloperGuide.md

### PR DESCRIPTION
**Pull request by nayato**
Reworked cloud to device scenarios adding QoS 2 support description and introducing RequestAckPairProcessor.

### CODE REVIEW

**Comment by nayato** [comment]
@nberdy, can you take a look?

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
-7. `MqttIotHubAdapter` peeks at the top message from the ACK-pending queue, and if it does not match the received PUBACK packet then `MqttIotHubAdapter` will stop processing (in this case skip following steps).
-8. `MqttIotHubAdapter` dequeues the message from the ACK-pending queue and completes the message on IoT Hub.
+4. If QoS=0,
+    - `MqttIotHubAdapter` sends a PUBLISH packet through the pipeline (see steps 13-16 above for CONNECT processing)
+    - At the same time `MqttIotHubAdater` completes the message on IoT hub.
+5. If QoS=1,
+    - `MqttIotHubAdapter` posts a PUBLISH packet to `RequestAckPairProcessor` responsible for handling PUBLISH-PUBACK communication.
+    - `RequestAckPairProcessor` sends PUBLISH packet through the pipeline (see steps 13-16 above for CONNECT processing).
+    - at the same time `RequestAckPairProcessor` enqueues an entry with the message state (packet identifier) in its internal ACK-pending queue. The ACK-pending queue is used for tracking messages in-flight that await on acknowledgement message to arrive from the device.
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Please capitalize the first letter and also add a comma after "same time" at the beginning.

Is the (packet identifier) the same as the message state?  That's what's implied by the parenthetical, and I want to verify that's what's meant.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - Eventually (assuming normal operation), PUBACK packet corresponding to the PUBLISH packet is received from the device and gets through the pipeline to `MqttIotHubAdapter`.
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Please say "the PUBACK packet corresponding..."

Also change "gets through the pipeline" to "goes through the pipeline"

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - `MqttIotHubAdapter` dispatches PUBACK packet to `RequestAckPairProcessor`.
+    - `RequestAckPairProcessor` peeks at the top message from its ACK-pending queue, and if it does not match the received PUBACK packet, PUBACK packet gets discarded and processing stops (following steps are skipped).
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Should be "...packet, the PUBACK packet gets discarded and processing stops (the following steps are skipped)."

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - `RequestAckPairProcessor` dequeues the message from the ACK-pending queue and calls `MqttIotHubAdapter` to complete acknowledgement.
+    - `MqttIotHubAdapter` completes the message on IoT hub.
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: What does "completes the message" mean?

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+5. If QoS=2,
+    - `MqttIotHubAdapter` posts a PUBLISH packet to `RequestAckPairProcessor` responsible for handling PUBLISH-PUBREC communication.
+    - `RequestAckPairProcessor` sends PUBLISH packet through the pipeline (see steps 13-16 above for CONNECT processing).
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Should be "sends a PUBLISH packet"

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - at the same time `RequestAckPairProcessor` enqueues an entry with the message state (packet identifier) in its internal ACK-pending queue.
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Please capitalize the first letter.

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - Eventually (assuming normal operation), PUBREC packet corresponding to the PUBLISH packet is received from the device and gets through the pipeline to `MqttIotHubAdapter`.
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Should be "..., the PUBREC packet corresponding..."

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - `MqttIotHubAdapter` dispatches PUBREC packet to `RequestAckPairProcessor`.
+    - `RequestAckPairProcessor` peeks at the top message from its ACK-pending queue, and if it does not match the received PUBREC packet, PUBREC packet gets discarded and processing stops (following steps are skipped).
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Should be "..., the PUBREC packet gets discarded..." (missing "the")

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - `RequestAckPairProcessor` dequeues the message from the ACK-pending queue and calls `MqttIotHubAdapter` to complete acknowledgement.
+    - `MqttIotHubAdapter` calls `IQos2StatePersistenceProvider.SetMessageAsync` to persist the fact of acknowledged PUBLISH packet delivery.
+    - `MqttIotHubAdapter` posts a PUBREL packet to `RequestAckPairProcessor` responsible for handling PUBREL-PUBCOMP communication.
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Should be "...packet to the `RequestAckPairProcessor`..." (missing "the")

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - `RequestAckPairProcessor` sends PUBREL packet through the pipeline (see steps 13-16 above for CONNECT processing).
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Should be "...sends the PUBREL packet..." (missing 'the')

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - at the same time `RequestAckPairProcessor` enqueues an entry with the message state in its internal ACK-pending queue.
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Please capitalize the first word.  Also there's a missing article, so it should be "At the same time the `RequestAckPairProcessor`..."

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - Eventually (assuming normal operation), PUBCOMP packet corresponding to the PUBREL packet is received from the device and gets through the pipeline to `MqttIotHubAdapter`.
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Should be "..., the PUBCOMP packet..."

Also please change "gets" to "goes"

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - `MqttIotHubAdapter` dispatches PUBCOMP packet to `RequestAckPairProcessor`.
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Missing 'the', should be "...displatches the PUBCOMP..."

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - `RequestAckPairProcessor` peeks at the top message from its ACK-pending queue, and if it does not match the received PUBCOMP packet, PUBCOMP packet gets discarded and processing stops (following steps are skipped).
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Should be, "...and if the message does not match the received PUBCOMP packet, the PUBCOMP packet gets discarded and processing stops (the following steps are skipped)."  Minor rewording and also added some articles.

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - `RequestAckPairProcessor` dequeues the message from the ACK-pending queue and calls `MqttIotHubAdapter` to complete acknowledgement.
+    - `MqttIotHubAdapter` completes message on IoT hub
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Not sure what completing a message means, but it should say "completes the message" (missing 'the')

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (nayato) [inline docs/DeveloperGuide.md]: message state means one of two things: src/ProtocolGateway.Core/Mqtt/AckPendingMessageState.cs (for PUBACK and PUBREC) or src/ProtocolGateway.Core/Mqtt/CompletionPendingMessageState.cs (for PUBCOMP). Not sure if it makes sense to cover data structures in scenarios. IMHO for people who would actually read through this there is enough pointers in a form of component class names.

Comment (nberdy) [inline docs/DeveloperGuide.md]: If packet identifier and message state are the same, then I have no further complaint with this line.

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (nayato) [inline docs/DeveloperGuide.md]: changed to "deletes the message from IoT hub

**Comment by nayato** [comment]
@nberdy, can you take another look?

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
 6. `MqttIotHubAdapter` adds the parsed variables as message properties and sends it to IoT Hub through `IDeviceClient`, which was acquired during connection establishement in step 12 above.
 7. If the QoS flag of the PUBLISH packet is set to 1 ("at least once"), `MqttIotHubAdapter` will send a PUBACK packet through the channel pipeline, which is similar to steps 13-16 above to process a CONNECT packet.
 
-### Protocol gateway receives message from IoT Hub (QoS 1) 
-Once `MqttIotHubAdapter` receives a messages from IoT Hub at any time after it has opened a receive link to IoT Hub:
+### Protocol gateway receives message from IoT Hub 
+Once `MqttIotHubAdapter` has opened a receive link to IoT hub, it can receive a message at any time. When message is received:
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Missing a word, should be "When a message is received"

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
-7. `MqttIotHubAdapter` peeks at the top message from the ACK-pending queue, and if it does not match the received PUBACK packet then `MqttIotHubAdapter` will stop processing (in this case skip following steps).
-8. `MqttIotHubAdapter` dequeues the message from the ACK-pending queue and completes the message on IoT Hub.
+4. If QoS=0,
+    - `MqttIotHubAdapter` sends a PUBLISH packet through the pipeline (see steps 13-16 above for CONNECT processing)
+    - At the same time `MqttIotHubAdater` deletes the message from IoT hub.
+5. If QoS=1,
+    - `MqttIotHubAdapter` posts a PUBLISH packet to `RequestAckPairProcessor` responsible for handling PUBLISH-PUBACK communication.
+    - `RequestAckPairProcessor` sends PUBLISH packet through the pipeline (see steps 13-16 above for CONNECT processing).
+    - At the same time `RequestAckPairProcessor` enqueues an entry with the message state in its internal ACK-pending queue. The ACK-pending queue is used for tracking messages in-flight that await on acknowledgement message to arrive from the device.
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Missing a word, should be "...that await on an acknowledgement message..."

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
...
+    - `MqttIotHubAdapter` dispatches PUBACK packet to `RequestAckPairProcessor`.
+    - `RequestAckPairProcessor` peeks at the top message from its ACK-pending queue, and if it does not match the received PUBACK packet, the PUBACK packet gets discarded and processing stops (the following steps are skipped).
+    - `RequestAckPairProcessor` dequeues the message from the ACK-pending queue and calls `MqttIotHubAdapter` to complete acknowledgement.
+    - `MqttIotHubAdapter` deletes the message from IoT hub.
+5. If QoS=2,
+    - `MqttIotHubAdapter` posts a PUBLISH packet to `RequestAckPairProcessor` responsible for handling PUBLISH-PUBREC communication.
+    - `RequestAckPairProcessor` sends a PUBLISH packet through the pipeline (see steps 13-16 above for CONNECT processing).
+    - At the same time the `RequestAckPairProcessor` enqueues an entry with the message state in its internal ACK-pending queue.
+    - Eventually (assuming normal operation), the PUBREC packet corresponding to the PUBLISH packet is received from the device and gets through the pipeline to `MqttIotHubAdapter`.
+    - `MqttIotHubAdapter` dispatches PUBREC packet to `RequestAckPairProcessor`.
```
Comment (nberdy) [inline docs/DeveloperGuide.md]: Missing a word, should be "dispatches a PUBREC packet"

**Comment by nberdy** [comment]
Made 3 comments, otherwise looks good.

**Pull request merged**